students/Thai_H/lesson6/mail_room_part4.py

```diff
@@ -0,0 +1,222 @@
+
+#----------------------------------------
+def sum_donation(d_amt):
+    return sum(d_amt)
+
+
+#-------------------------------------------
+def num_of_donation(n_donation):
+    return len(n_donation)
+
+
+#------------------------------------------
+def avg_donation(total, nd):
+    if nd == 0:
+        raise ValueError('can not divide by zero')
+    return total / nd
+
+
+#--------------------------------------------------------------------------
+def sort_donor_list(unsorted_list):
+    sorted_donation_list = []
+    for d in unsorted_list:
+        total_donation =  sum_donation(unsorted_list[d])
+        number_of_donation = num_of_donation(unsorted_list[d])
+        average = avg_donation(total_donation,  number_of_donation)
+        sorted_donation_list.append([d, total_donation, number_of_donation, average])
+
+    sorted_donation_list = sorted(sorted_donation_list, key = itemgetter(1), reverse = True)
+    return sorted_donation_list
+
+#--------------------------------------------------------------------------
+def sort_by_total_donation(sorted_donation_list):
+    return sorted_donation_list[1]
+
+
+#------------------------------------------------------------------------------
+def create_report():
+    sorted_list = sort_donor_list(list_of_donors)
+    report_header = ['Donor Name', 'Total Donation Amt', 'Num of Donation',  'Avg Per Donation']
+    print('\n\n{:<25}{:>20}{:>20}{:>20}'.format(*report_header))
+    print('-'*90)
+    for donor in sorted_list:
+        print('{:<26}${:>15,.2f}{:>15}{:>25,.2f}'.format(donor[0], donor[1], donor[2], donor[1] / donor[2]))
+    print('\n\n')
+
+
+#------------------------------------------------------------------------------
+def test_create_report():
+    for key, item in list_of_donors.items():
+        print("{:<20}: {}".format(key, item))
+
+
+#------------------------------------------------------------------------------
+def init_donor_list():
+    # create initial list of donors and their donations
+    global list_of_donors
+    list_of_donors = {'William Boeing': [12.86, 2.00, 3.00, 4.00, 5.00],
+                      'Steve Jobs':     [6.00, 7.00, 8.00],
+                      'Paul Allen':     [9.00, 10.00, 11.00],
+                      'Charles Flint':  [12.00, 13.00, 14.00],
+                      'Thomas Edison':  [15.00, 16.00, 17.00]  }
+
+
+#------------------------------------------------------------------------------
+def main_menu_selection():
+    response = input('\n\nMAIN MENU\nChoose from one of the options below:\n\
+        T) THANK-YOU MENU\n\
+        R) Create a Report\n\
+        Q) Quit the program\n\
+        Please type T, R, or Q: ')
+    return response
+
+
+#------------------------------------------------------------------------------
+def quit_program():
+    print('Bye!')
+    sys.exit()
+
+
+#-----------------------------------------------------------------------------
+def invalid_menu():
+    print('\nNot a valid option!\n\n')
+    main()
```

Comment:
you're in a while True loop.. main() will resume running on it's own - avoid creating additional nested loops